src/Storages/MergeTree/MergeTreeBlockReadUtils.h

```diff
@@ -64,6 +64,13 @@ struct MergeTreeBlockSizePredictor
     /// Predicts what number of rows should be read to exhaust byte quota per column
     size_t estimateNumRowsForMaxSizeColumn(size_t bytes_quota) const
     {
+        /// Below we calculate the following expression: `bytes_quota / max_size_per_row` and then cast its result to size_t.
+        /// The problem is that the conversion to double (`max_size_per_row` is double) can overflow size_t if `bytes_quota` is close to `SIZE_MAX`
+        /// and an attempt to convert this value back to size_t will result in undefined behavior.
+        /// ATST, if `bytes_quota` is close to `SIZE_MAX`, we can safely ignore it and assume that no limit was set.
+        if (static_cast<double>(bytes_quota) >= static_cast<double>(std::numeric_limits<size_t>::max()))
+            return 0;
```

Comment:
(I'm more in the camp of throwing an exception because the user specified such a large value probably unintentionally but returning 0 is also okay.)